Disable diagnostic service for projects that we know we don't have all information to provide good results.

heejaechang opened the pull request:
this should make roslyn to turn off full project diagnostics when intellisense build is failed for the project.

when we detect intellisense build failure, VS should put actionable data in error list detail pane.
also bing search for the issue should work as well if one prefer blogs.

this change also fix some issue on cross language p2p references. before there was case where we fail to connect cross language p2p reference causing a lot of errors.

this change also include bulk diagnostic update events improvement. this should let us to deal better (especially when we clean those up) when there are a lot of diagnostics.

devdiv side ask mode PR.
https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS/pullrequest/14158?view=discussion

both github and devdiv side needed to be checked in to fix all related issues.

...

related PR in master - https://github.com/dotnet/roslyn/pull/8878

heejaechang commented:
with this fix - https://github.com/dotnet/roslyn/issues/6809 shows no errors.

with this fix - https://github.com/dotnet/roslyn/issues/6779 - this kind of cross language issues should be resolved. 
(confirmed it with https://github.com/dotnet/roslyn/issues/6809 but I dont have repro for https://github.com/dotnet/roslyn/issues/6779 but assuming from the symptom described, most likely same issue)

MattGertz commented:
Approved pending code review signoff.  (I can't get to the devdiv site at the moment.)

heejaechang commented:
ping? I need sign off.

jasonmalinowski commented on src/EditorFeatures/Test/Diagnostics/DiagnosticAnalyzerServiceTests.cs:
```diff
@@ -0,0 +1,132 @@
+            // listen to events
+            service.DiagnosticsUpdated += (s, a) => Assert.Empty(a.Diagnostics);
+
+            // now call each analyze method. none of them should run.
```
How does this test assert none of them run?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -455,13 +457,50 @@ protected int AddMetadataReferenceAndTryConvertingToProjectReferenceIfPossible(s
+            //    2. output path of referenced cross language project might be changed to right one 
+            //       once it is already added as a metadata reference.
+            //
+            // but this has one concequence. even if a user adds a project in the solution as 
```
"consequence"

heejaechang replied:
done.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -455,13 +457,50 @@ protected int AddMetadataReferenceAndTryConvertingToProjectReferenceIfPossible(s
+            // as a metadata reference.
             AddMetadataReferenceCore(this.MetadataReferenceProvider.CreateMetadataReference(this, filePath, properties));
 
+            // here, we change behavior compared to old LS. regardless of file being exist or not, 
```
We should avoid acronyms for others who want to read this.

heejaechang replied:
done

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -455,13 +457,50 @@ protected int AddMetadataReferenceAndTryConvertingToProjectReferenceIfPossible(s
+            //
+            // after this point, Roslyn will take care of non-exist metadata reference.
+            //
+            // BUT, this doesn't sovle the issue where actual metadata reference 
```
typo.

heejaechang replied:
done

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_IIntellisenseBuildTarget.cs:
```diff
@@ -0,0 +1,134 @@
+
+        public void SetIntellisenseBuildResult(bool succeeded, string reason)
+        {
+
```
Blank line.

heejaechang replied:
done

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_IIntellisenseBuildTarget.cs:
```diff
@@ -0,0 +1,134 @@
+                return null;
+            }
+
+            var fileLogger = buildParameters.Loggers?.OfType<FileLogger>()?.FirstOrDefault();
```
I'm not convinced we couldn't have multiple FileLoggers, In any case, this entire function seems supremely dangerous. I would strongly prefer we remove the whole thing and just write up a good blog post explaining how to find this.

heejaechang replied:
@jasonmalinowski well, the good blog post still can't let you know the exact log file name. that is the problem. only way is you go through all designtime.log files in the directory, and open one by one or do some kind of file in directory search to find what you want. all just extra hurdle for users to get to information they need.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_IIntellisenseBuildTarget.cs:
```diff
@@ -0,0 +1,134 @@
+        private static readonly FieldInfo s_fieldInfo = typeof(BuildManager).GetField("_buildParameters", BindingFlags.GetField | BindingFlags.Instance | BindingFlags.NonPublic);
+
+        // REVIEW: can temp path changed while process running?
+        private static readonly string GeneralLogFile = $"{Path.GetTempPath()}\\<guid>.designtime.log";
```
I don't like that we're duplicating this here, especially since I'd like to make changes to where it writes in the project system so it's easier to find. Can we just not do this?

heejaechang replied:
are we going to change BuildManager?

heejaechang replied:
I think we have 3 options.

1. just take out reflection until we have better way to do it. (less optimal user experience)
2. take reflection for update 2 and create tracking issue for it for update 3.
3. make buildmanageraccessor change in update 2.

I would prefer option (2). but I am not decision maker, so I will let leads to decide.

the reason I don't like (1) is that, according to my own experience, without log name, experience is just bad. it is like windows tell me something failed with error code XXX. which require me to do more work to resolve the issue.

for (3), I dont know that part of code, it is first time for me to take a look at that code. so not sure whether I am a right person to make right change to that code to make it exposed.

heejaechang replied:
opened an issue to add proper API for logs.

https://github.com/dotnet/roslyn/issues/9158

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/IVisualStudioWorkspaceHost2.cs:
```diff
@@ -0,0 +1,14 @@
+    // TODO: Find a better name for this
+    internal interface IVisualStudioWorkspaceHost2
+    {
+        void OnIsComplete(ProjectId projectId, bool succeeded);
```
Parameter should probably be called "complete" or something, not "succeeded".

heejaechang replied:
changed to HasAllInformation

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/Interop/IIntellisenseBuildTarget.cs:
```diff
@@ -0,0 +1,16 @@
+    [Guid("F304ABA7-2448-4EE9-A706-D1838F200398")]
+    internal interface IIntellisenseBuildTarget
+    {
+        // currently, reason is not being used.
```
Aren't you using consuming it?

heejaechang replied:
current "reason" is not set by project system. I can remove it if we don't see it ever being used.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.HostProject.cs:
```diff
@@ -60,6 +60,8 @@ public ProjectInfo CreateProjectInfoForCurrentState()
                     isSubmission: false,
                     hostObjectType: null);
+
+                return info.WithIsComplete(isComplete: false);
```
You didn't mean true? MiscellaneousFiles to me is the canonical example of this should be set to true, since we know semantics are going to be busted.

heejaechang replied:
I think we already talked about it in the other PR.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:
```diff
@@ -1040,6 +1040,19 @@ internal void OnAdditionalDocumentTextUpdatedOnDisk(DocumentId documentId)
+        internal void OnIsComplete(ProjectId projectId, bool succeeded)
+        {
+            // this can't be null
+            var state = CurrentSolution.GetProjectState(projectId);
```
Why is this still groveling into ProjectState, which should be internal to the core Workspace type?

jasonmalinowski replied:
And shouldn't this just be in the core workspaces layer?

heejaechang replied:
sure, since it is in ProjectInfo. probably make sense to push it down to core workspace. will do.

jasonmalinowski commented on src/VisualStudio/Core/Def/ServicesVSResources.resx:
```diff
@@ -513,4 +513,28 @@ Use the dropdown to view and switch to other projects this file may belong to.</
+    <value>Project loading failed.</value>
+  </data>
+  <data name="IntellisenseBuildFailedDescription" xml:space="preserve">
+    <value>To see what caused the issue, please try below.
```
Once again, I would strongly prefer just a fwlink to a page we can write up, with proper instructions. If for no other reason that step 1 in your instructions means they can no longer read the rest of the instructions. :smile:

heejaechang replied:
not sure what you mean? anyway, providing direct link is not hard. we can set help link to that direct link in diagnostic. both can be provided.

jasonmalinowski replied:
Just a discovery nit: if people see this, will they know they can click the link to get to a full page?

heejaechang replied:
that is behavior of all other errors. so I believe they will know.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:
```diff
@@ -91,195 +62,100 @@ public string OutputFilePath
+        /// <summary>
+        /// True if project information is complete. <see cref="ProjectInfo.IsComplete"/> for more info.
+        /// 
+        /// this is for internal usage only. for public usage, use <see cref="IsCompleteAsync(CancellationToken)"/>
```
Why would internal be different here? This seems really confusing to me.

heejaechang replied:
I removed the comment and change name to be different than the other. I wrote it since IsComplete in ProjectInfo and IsCompleteAsync Project uses similar name when two are actually different thing.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:
```diff
@@ -390,12 +266,21 @@ public Task<Compilation> GetCompilationAsync(CancellationToken cancellationToken
         /// <summary>
-        /// Determines if the compilation returned by <see cref="GetCompilationAsync"/> has all the references it's expected to have.
+        /// Determines if the compilation returned by <see cref="GetCompilationAsync"/> has full information.
+        /// see <see cref="ProjectInfo.IsComplete"/> for more info.
```
Capitalize 's'.

heejaechang replied:
comment removed.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:
```diff
@@ -390,12 +266,21 @@ public Task<Compilation> GetCompilationAsync(CancellationToken cancellationToken
+
+        /// <summary>
+        /// Determines if the compilation returned by <see cref="GetCompilationAsync"/> and all its referenced projects have full information.
+        /// see <see cref="IsCompleteAsync(CancellationToken)"/> and <see cref="ProjectInfo.IsComplete"/> for more info.
```
Capitalize 's'.

heejaechang replied:
comment removed

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:
```diff
@@ -390,12 +266,21 @@ public Task<Compilation> GetCompilationAsync(CancellationToken cancellationToken
+        /// Determines if the compilation returned by <see cref="GetCompilationAsync"/> and all its referenced projects have full information.
+        /// see <see cref="IsCompleteAsync(CancellationToken)"/> and <see cref="ProjectInfo.IsComplete"/> for more info.
         /// </summary>
-        // TODO: make this public
```
TODO shouldn't be deleted, since this is still not public?

heejaechang replied:
ah, ya, it went to public and came back to internal :)

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs:
```diff
@@ -25,7 +25,9 @@ internal partial class ProjectState
         private readonly AsyncLazy<VersionStamp> _lazyLatestDocumentTopLevelChangeVersion;
-        private AnalyzerOptions _analyzerOptions;
+
+        // this will be initialized lazily.
```
Why not just use Lazy<>? Also, should this be XML doc comments so it can point to where it's initialized?

heejaechang replied:
that I dont know, I just changed name. changing behavior should be consulted by the one who added it I believe. anyway, not dealing with it in this PR.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs:
```diff
@@ -602,8 +590,16 @@ public CompilationInfo(Compilation compilation, bool hasCompleteReferences)
+
                     var newReferences = new List<MetadataReference>();
+
+                    // we don't consider metadata reference being actually exist on disk as a part of complete reference
```
I don't get this statement. ResolveAssemblyReferences has already filtered out references that don't exist on disk that are user-added. The only references we might find missing here are P2P ones where we most definitely might count that.

heejaechang replied:
well, user can delete it once added.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs:
```diff
@@ -660,6 +657,30 @@ public CompilationInfo(Compilation compilation, bool hasCompleteReferences)
+                    return false;
+                }
+
+                foreach (var projectReference in this.ProjectState.ProjectReferences)
```
Given we're already doing this in the other function which is also computing compute bits, why not just do it in the same loop?

heejaechang replied:
the other loop is not done in a way getting compilation is straightfoward, it pass down projects and at some point it create compilation. this is way simpler.

jasonmalinowski commented on src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs:
```diff
@@ -1450,8 +1450,81 @@ public void TestFrozenPartialProjectAlwaysIsIncomplete()
+                    LanguageNames.VisualBasic,
+                    projectReferences: new[] { new ProjectReference(dependsOnVbProject.Id) }));
+
+            Assert.False(csBrokenProject.IsCompleteAsync().Result);
```
Can we break this apart into a few tests that more clearly state what they're asserting? This melts my brain.

heejaechang replied:
sure.

jasonmalinowski commented:
Is it also possible to (without too much work) break apart the commit into a few pieces? A +1,000/-500 diff commit is going to be a headache to work through if ever we have to git bisect to this.

jasonmalinowski commented:
I've reviewed the project/workspace half of it, but can't speak at all to the diagnostics half of it. I still have some reservations or concerns I'd like to see addressed before this goes in.

heejaechang commented:
@jasonmalinowski not sure about how to break the comments? can you give me some example?

heejaechang commented:
@CyrusNajmabadi I addressed your comment in this PR since PR is in both update 2 and master. I will port this change back to master and drop the master PR.

heejaechang commented:
@jasonmalinowski @CyrusNajmabadi I found a new issue where cross language reference won't work which only way to fix seems tracking metadata reference missing.

can one of you take a look at this commit? 
https://github.com/heejaechang/roslyn/commit/9b66f2d9ff7baa9c64a3907d431f7eb1434e63bb

heejaechang commented:
@srivatsn @mavasani I really need code review for analyzer side.

mavasani commented on src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateSet.cs:
```diff
@@ -39,10 +39,18 @@ public DiagnosticState GetState(StateType stateType)
             }
 
-            public void Remove(object documentOrProjectId)
+            public void Remove(object documentOrProjectId, bool includeProjectState = true)
```
Probably easier to understand if the flag is called "onlyDocumentStates" and the check below is negated.

heejaechang replied:
sure

heejaechang replied:
done

mavasani commented on src/Features/Core/Portable/Diagnostics/BaseDiagnosticIncrementalAnalyzer.cs:
```diff
@@ -232,7 +235,7 @@ public virtual void LogAnalyzerCountSummary()
         // internal for testing purposes.
         internal Action<Exception, DiagnosticAnalyzer, Diagnostic> GetOnAnalyzerException(ProjectId projectId)
         {
-            return Owner?.GetOnAnalyzerException(projectId, DiagnosticLogAggregator);
```
Do we know when this could have ever been null?

heejaechang replied:
not sure, but when I checked, it never be null now.

mavasani commented on src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService_UpdateSource.cs:
```diff
@@ -42,17 +42,27 @@ private DiagnosticAnalyzerService(IDiagnosticUpdateSourceRegistrationService reg
+            }
+        }
+
+        internal void RaiseBulkDiagnosticsUpdated(Action<Action<DiagnosticsUpdatedArgs>> eventAction)
```
Can you add a comment that this method is added to avoid spawning multiple tasks?

heejaechang replied:
sure.

heejaechang replied:
done

mavasani commented on src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.cs:
```diff
@@ -392,24 +428,20 @@ internal CompilationWithAnalyzers GetCompilationWithAnalyzers(Project project, C
         {
-            if (Owner.IsAnalyzerSuppressed(stateSet.Analyzer, userDiagnosticDriver.Project))
+            // this project is not in a good state. only continue if it is opened document.
+            if (!await project.HasSuccessfullyLoadedAsync(cancellationToken).ConfigureAwait(false) && !openedDocument)
```
What happens when project had loaded successfully, had diagnostics and then goes into this bad state. Shouldn't we clear the diagnostics and then skip running analyzers?

heejaechang replied:
that is what it does.

if (await SkipRunningAnalyzerAsync(document.Project, stateSet.Analyzer, openedDocument, skipClosedFileCheck: false, cancellationToken: cancellationToken).ConfigureAwait(false))
                    {
                        await ClearExistingDiagnostics(document, stateSet, StateType.Syntax, cancellationToken).ConfigureAwait(false);
                        continue;
                    }

mavasani commented on src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs:
```diff
@@ -497,39 +538,61 @@ public ProjectState WithProjectReferences(IEnumerable<ProjectReference> projectR
+            return reference != null && !PortableShim.File.Exists(reference.FilePath);
+        }
+
+        private static bool ComputeMissingMetadataReference(IEnumerable<PortableExecutableReference> references)
```
Should be called HasMissingMetadataReference?

heejaechang replied:
there is already property with that name. so used Compute instead.

mavasani commented:
Diagnostic service changes LGTM. I am hoping this resolves a lot of OOMs being reported.

jasonmalinowski commented:
For the second commit that was recently added, why is an approach like https://github.com/jasonmalinowski/roslyn/commit/9fa370a9b8c67f6ac51fcd836c06b890f78b4a0f not sufficient? Not only is churning a _lot_ less code, but it doesn't require a bunch of tracking of state, and also makes it more obvious that things like file watcher changes work, which I'm not convinced the current code does.

jasonmalinowski commented:
And to explain my concern with the new commit: I don't understand how the project system is keeping everything in sync with what might _actually_ be determined once we have a compilation. Consider scenarios where the file exists and is then deleted shortly afterwards, or the file exists but we can't read it, etc. Going to the compilation seems much, much safer.

heejaechang commented:
my concern was perf since it is one of hot path. another thing is I would like to report diagnostics when there is missing reference so that people will know about it.

heejaechang commented:
also, if we go that route, we don't even need Compilation, we can just call reference.GetMetadata/errors to see whether metadata exist.

jasonmalinowski commented:
@heejaechang: If that works then yes, that's even simpler. My biggest concern is ensuring snapshot stability -- i.e. if I have a compilation that has those metadata references, and I also call this API, I shouldn't get inconsistent results. I believe your implementation does risk that. My understanding is we ensure in Visual Studio that once a call to GetMetadata() for our implementation of metadata threw, it continues to throw or not throw. Once the file is added or deleted, we recreate the instance and reload it in the workspace. Thus, you get all the file watcher stuff for free.

As far as the performance concerns, I don't understand them. Fetching the metadata is going to happen as soon as _anybody_ asks for any symbol, which is pretty much any feature. We already have caches left and right to ensure that's fast. Furthermore, none of the work is "wasted" in the sense that [a] if we say everything is A-OK, we're going to run analyzers (which will fetch metadata and symbols), or [b] we say something is broken, but still every IDE feature is using that same compilation to do completion, colorization, etc, etc. We might be moving CPU around, but I don't think we're creating work where we previously didn't.

> we don't even need Compilations

Technically, no. But if the function returns "true" that everything is complete, the immediate next step fo the caller is going to be to ask for a compilation. So I'm not sure we save anything in practice.

jasonmalinowski commented:
> another thing is I would like to report diagnostics when there is missing reference so that people will know about it.

Why wouldn't the compiler already report that itself as any other error would? If the diagnostics engine needs to report it for some reason, then it can do a check similar to what I'm proposing, and then it's a local concern vs. a bunch of API changes for pushing stuff around.

heejaechang commented:
not for cross language reference. if it was not for cross language reference, I dont think we even need this.

heejaechang commented:
addressed all feedbacks. now log uses project name and target name when saving. those changes are in https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS/pullrequest/14158?view=files

heejaechang commented:
ping?

CyrusNajmabadi commented:
The workspace side looks a ton better.  One question: can you point me at the code that now checks these new 'HasFullyLoaded' results and uses that to change what diagnostics are shown to hte user?

Thanks!

heejaechang commented:
here. https://github.com/dotnet/roslyn/pull/9017/files#diff-78e5a2e9f41571f2071673de09b9f57dR433

heejaechang commented:
@jasonmalinowski ping? thank you!

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -455,13 +457,50 @@ protected int AddMetadataReferenceAndTryConvertingToProjectReferenceIfPossible(s
+            // But, this doesn't sovle the issue where actual metadata reference 
+            // (not cross language p2p reference) is missing at the time project is opened.
+            //
+            // in that case, msbuild itself filter those references out, so project system doesn't know 
```
Just a documentation point: MSBuild doesn't filter P2P references if they're missing, since those go through AssemblyFiles to RAR instead of the Assemblies input. But there's enough project system stuff still filtering it.

heejaechang replied:
will change doc, thank you.

jasonmalinowski commented on src/VisualStudio/Core/Def/ServicesVSResources.resx:
```diff
@@ -537,4 +537,23 @@ Use the dropdown to view and switch to other projects this file may belong to.</
+    <value>To see what caused the issue, please try below.
+
+1. Close VS
+2. Set environment variable (set TraceDesignTime=true)
```
Reorder 2 and 3, perhaps, and then this can be "open a Visual Studio Developer Command Prompt, and then run set TraceDesignTime=true"? Then it makes 4 easier to explain.

heejaechang replied:
okay.

jasonmalinowski commented on src/VisualStudio/Core/Def/ServicesVSResources.resx:
```diff
@@ -537,4 +537,23 @@ Use the dropdown to view and switch to other projects this file may belong to.</
+1. Close VS
+2. Set environment variable (set TraceDesignTime=true)
+3. Delete .vs directory/.suo file
+4. Restart VS from the VS developer command prompt you set the environment varaible (devenv)
```
Typo in variable, also this isn't really a complete sentence.

heejaechang replied:
so, what should I put?

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs:
```diff
@@ -660,6 +648,43 @@ public CompilationInfo(Compilation compilation, bool hasCompleteReferences)
+            {
+                foreach (var reference in metadataReferences)
+                {
+                    if (compilation.GetAssemblyOrModuleSymbol(reference) == null)
```
I assume you've verified once this actually works? I admit I wrote the code in my branch but didn't test it.

heejaechang replied:
yep. tested it.

jasonmalinowski commented:
:+1: on the workspaces/project system half, which is the only part I feel qualified to review. A few wording nits for comments and some UI displayable text, but assuming this all tests out fine I have no functional concerns.

heejaechang commented:
@jasonmalinowski thank you!
